File: wltbase/lifecycle.go
package wltbase

import (
	"context"

	"github.com/KarpelesLab/libwallet/wltintf"
	"github.com/KarpelesLab/pobj"
)

func init() {
	pobj.RegisterStatic("Lifecycle:update", lifecycleUpdate)
}

// lifecycleUpdate receives the host's current lifecycle status
// (`foreground` / `background` / `paused` / `resumed` / …). It echoes
// the accepted status and drives pause/resume of background work:
// the balance poller is paused in background, resumed on foreground.
func lifecycleUpdate(ctx context.Context, in struct {
	Status string `json:"status"`
}) (any, error) {
	if e, ok := wltintf.GetEnv(ctx).(*env); ok && e.poller != nil {
		switch in.Status {
		case "background", "paused":
			e.poller.pause()
		case "foreground", "resumed", "active", "":
			e.poller.resume()
		}
	}
	return map[string]any{"status": in.Status}, nil
}

File: rust/src/handlers/lifecycle.rs
//! `Lifecycle:update` (port of wltbase/lifecycle.go). The host reports its app
//! lifecycle state so the library can pause/resume background work.

use serde_json::{json, Value};

use crate::Env;

use super::ApiResult;

/// `Lifecycle:update` {Status} — accept the host lifecycle status and echo it.
/// In Go this pauses the balance poller in `background`/`paused` and resumes it
/// on `foreground`/`resumed`/`active`; the poller isn't ported yet, so the
/// pause/resume is a no-op and only the echo contract is honored.
pub fn update(_env: &Env, params: &Value) -> ApiResult {
    let status = params.get("Status").or_else(|| params.get("status")).and_then(Value::as_str).unwrap_or("");
    Ok(json!({ "status": status }))
}
